### PR TITLE
Add a way to propagate test failure in aftereach block to the main UI 

### PR DESCRIPTION
**Pull request by jiujiu1123**

### CODE REVIEW

**Comment by jiujiu1123** [comment]
This basically allows recordIterationComplete to be called multiple times and each time it would write out a new test execution report. a little bit more IO but it shouldn't be too bad.

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java [hunk 1]
```diff
@@ -22,8 +22,6 @@
 public class TestExecutionReport {
     private static final Logger logger = Logger.getLogger(TestExecutionReport.class.getName());
 
-    private boolean hasReportBeenMaterialized = false;
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:25]: Why did you remove this?

File: src/main/java/cloud/filibuster/junit/server/core/FilibusterCore.java [hunk 2]
```diff
@@ -326,28 +326,24 @@ public synchronized boolean isNewTestExecution(String serviceName) {
...
     // JUnit hooks.
 
+
     // This is an old callback used to exit the Python server with code = 1 or code = 0 upon failure.
     public synchronized void completeIteration(int currentIteration) {
-        logger.info("[FILIBUSTER-CORE]: completeIteration called, currentIteration: " + currentIteration);
-
-        if (currentConcreteTestExecution != null) {
-            currentConcreteTestExecution.printRPCs();
-            currentConcreteTestExecution.writeTestExecutionReport(currentIteration, /* exceptionOccurred= */ false, /* throwable= */ null);
```
Comment (cmeiklejohn) [inline src/main/java/cloud/filibuster/junit/server/core/FilibusterCore.java:335]: Where is this write method being called now?  I can't find it.

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java [hunk 3]
```diff
@@ -250,43 +248,45 @@ public void writeTestReport(int currentIteration, boolean exceptionOccurred, @Nu
...
         if (throwable != null) {
-            assertionFailureMessage = throwable.getMessage();
+            if (assertionFailureMessage == null || assertionFailureMessage.equals("")) {
+                assertionFailureMessage = throwable.getMessage();
+            } else {
+                assertionFailureMessage = assertionFailureMessage + "; " + throwable.getMessage();
+            }
         }
 
-        if (!hasReportBeenMaterialized) {
```
[the remark this answers is not in the document]
Comment (jiujiu1123) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:256]: Because I want everytime recordIterationComplete to be called for a new report to materialize and write out

**Comment by cmeiklejohn** [comment]
I'm concerned with the lack of tests for this.  

For example:

- Does this alway write out the report on failure/not on failure?
- Does this work with multiple after blocks? (one throws, all throw, vs. none throw.)
- Does this work with zero-after blocks? (one throws, all throw, vs. none throw.)
- etc.

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmeiklejohn) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:256]: Wouldn't it be better to write it out when the test finishes, store whether it's been materialized or not and pass vs. fail and then only write out again if the status changes?  Then, we aren't potentially writing it out for code n times if it has n after blocks?

File: src/main/java/cloud/filibuster/junit/server/core/FilibusterCore.java [hunk 4]
```diff
@@ -326,28 +326,24 @@ public synchronized boolean isNewTestExecution(String serviceName) {
...
         logger.info("[FILIBUSTER-CORE]: completeIteration called, currentIteration: " + currentIteration + ", exceptionOccurred: " + exceptionOccurred);
 
         if (currentConcreteTestExecution != null) {
-            currentConcreteTestExecution.printRPCs();
+            if (shouldPrintRPCSummary) {
+                currentConcreteTestExecution.printRPCs();
+            }
 
             if (exceptionOccurred != 0) {
                 currentConcreteTestExecution.writeTestExecutionReport(currentIteration, /* exceptionOccurred= */ exceptionOccurred != 0, /* throwable= */ throwable);
```
[the remark this answers is not in the document]
Comment (jiujiu1123) [reply src/main/java/cloud/filibuster/junit/server/core/FilibusterCore.java:349]: This is only for printing, so we are not inundating the user with console logs

**Comment by jiujiu1123** [comment]
@cmeiklejohn  lmk what you think

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java [hunk 5]
```diff
@@ -249,11 +277,21 @@ public void writeTestReport(int currentIteration, boolean exceptionOccurred, @Nu
...
         testExecutionPassed = !exceptionOccurred;
 
-        if (throwable != null) {
-            assertionFailureMessage = throwable.getMessage();
-        }
+        if (!hasReportBeenMaterialized || throwable != null) {
+            if (throwable != null) {
+                StringWriter sw = new StringWriter();
+                PrintWriter pw = new PrintWriter(sw);
+                throwable.printStackTrace(pw);
```
[the remark this answers is not in the document]
Comment (cmeiklejohn) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:284]: Do we need this print here?

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java (hunk 1 above)
Comment (cmeiklejohn) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:25]: We talked about you restoring this, right?

File: CHANGELOG [hunk 6]
```diff
@@ -1,3 +1,6 @@
+1.6.2:
```
Comment (cmeiklejohn) [inline CHANGELOG:1]: You're gonna have to merge master in and make this 1.6.3.

File: build.gradle [hunk 7]
```diff
@@ -207,7 +207,7 @@ javadoc {
 
 group = "cloud.filibuster"
 archivesBaseName = "instrumentation"
-version = "1.6.1"
+version = "1.6.2"
```
Comment (cmeiklejohn) [inline build.gradle:210]: You're gonna have to merge master in and make this 1.6.3.

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java (hunk 1 above)
Comment (jiujiu1123) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:25]: https://github.com/filibuster-testing/filibuster-java-instrumentation/blob/test-failure-showing/src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java#L280

should be restored

File: src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiujiu1123) [reply src/main/java/cloud/filibuster/junit/server/core/reports/TestExecutionReport.java:284]: Yes we do, this is not printing to console but to print it to the string so that it will be reflected in the test file

**Comment by cmeiklejohn** [comment]
Can you resolve the conflicts with the CHANGELOG and the build number?

**Comment by cmeiklejohn** [comment]
@jiujiu1123 one thing we need to do here is make this option configurable via:

- the annotation
- a system property.

Can you please make those changes?  If you need guidance, look at how suppress_combinations is also used in both a.) the annotation (which takes first precedence) and then the system property.  It should default to on.

**Comment by cmeiklejohn** [comment]
@jiujiu1123 disregard, I realized this is only reporting on the failure not controlling pass/fail, so you don't need the property/annotation.

**Review by cmeiklejohn — APPROVED** [review]
